tests/unit_tests.cpp

```diff
@@ -985,53 +985,114 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(explicit_buffer_copy_host_ptr, _dimensions,
+  std::vector<size_t> host_buf_window(buf_size.size(), canary);
+  const auto full_dim0_stride = buf_size[0];
+  const auto contiguous_dim0_stride = window_range[0];
+  const auto window_dim0_offset = window_offset[0];
+  for (size_t i = window_dim0_offset; i < window_dim0_offset + contiguous_dim0_stride; ++i) {
+    const auto full_dim1_stride = (d >= 2 ? buf_size[1] : 1);
+    const auto contiguous_dim1_stride = (d >= 2 ? window_range[1] : 1);
+    const auto window_dim1_offset = (d >= 2 ? window_offset[1] : 0);
+    for (size_t j = window_dim1_offset; j < window_dim1_offset + contiguous_dim1_stride; ++j) {
+      const auto full_dim2_stride = (d == 3 ? buf_size[2] : 1);
+      const auto contiguous_dim2_stride = (d == 3 ? window_range[2] : 1);
+      const auto window_dim2_stride = (d == 3 ? window_offset[2] : 0);
+      for (size_t k = window_dim2_stride; k < window_dim2_stride + contiguous_dim2_stride; ++k) {
+        const size_t id_in_window = i * full_dim1_stride * full_dim2_stride + j * full_dim2_stride + k;
+        const size_t id_in_contiguous = (i - window_offset[0]) * contiguous_dim1_stride * contiguous_dim2_stride
+            + (j - window_dim1_offset) * contiguous_dim2_stride
+            + (k - window_dim2_stride);
+        host_buf_contiguous[id_in_contiguous] = host_buf_full[id_in_window];
+        host_buf_window[id_in_window] = host_buf_full[id_in_window];
       }
     }
   }
 
   cl::sycl::queue queue;
-  cl::sycl::buffer<size_t, d> buf{buf_size};
 
+  // copy full buffer without strides
+  cl::sycl::buffer<size_t, d> device_buf_full{buf_size};
   queue.submit([&](cl::sycl::handler& cgh) {
-    auto acc = buf.template get_access<cl::sycl::access::mode::discard_write>(cgh);
-    cgh.copy(host_buf_source.data(), acc);
+    auto acc = device_buf_full.template get_access<cl::sycl::access::mode::discard_write>(cgh);
+    cgh.copy(host_buf_full.data(), acc);
   });
 
+  // copy contiguous buffer into window on the device buffer
+  cl::sycl::buffer<size_t, d> device_buf_window{buf_size};
   queue.submit([&](cl::sycl::handler& cgh) {
-    auto acc = buf.template get_access<cl::sycl::access::mode::read_write>(cgh);
-    cgh.parallel_for<kernel_name<class explicit_buffer_copy_host_ptr_mul3, d>>(
-      buf_size, [=](cl::sycl::item<d> item) {
-        acc[item] = acc[item] * 3;
-      });
+    auto acc = device_buf_window.template get_access<cl::sycl::access::mode::discard_write>(cgh);
+    cgh.fill(acc, canary);
+  });
+  queue.submit([&](cl::sycl::handler& cgh) {
+    auto acc = device_buf_window.template get_access<cl::sycl::access::mode::discard_write>(cgh,
+        window_range, window_offset);
+    const auto fake_shared_ptr = std::shared_ptr<size_t>(host_buf_contiguous.data(), [](size_t *){});
+    cgh.copy(fake_shared_ptr, acc);
   });
 
-  std::shared_ptr<size_t> host_buf_result(new size_t[buf_size.size()],
-    std::default_delete<size_t[]>());
+  // modify each buffer on the device to avoid potential transfer elision
```

Comment:
Is this really necessary? As soon as a host pointer is involved, I don't think that a SYCL implementation could elide an explicit copy operation as we cannot know what has happened to the host pointer. hipSYCL will also never elide explicit copies between accessors ("you asked explicitly for a copy, so you are getting one").

Reply:
Since up to the first synchronization point in line 1090 (`queue.wait`) the execution is asynchronous, the host memory cannot be modified from user code without causing a race with the async copy. I'd argue an implementation could potentially detect that this host -> device -> host round trip is unobservable and replace it with a memcpy. Not that I see any reason why hipSYCL should ever implement that kind of logic.

I don't feel strongly about this, if you think it's unnecessary I'll remove it. Your call.

Reply:
I also don't have strong feelings about this, but I'm leaning more towards removing. After all, these are the hipSYCL unit tests, so we can rely on guarantees that hipSYCL makes.

It's an interesting point you make. I think in order to observe that the host->device->host copy is unecessary the implementation would need to follow a lazy submission approach and evaluate the DAG only at the synchronization point. My understanding is that in SYCL 1.2.1 a `submit()` should always correspond to an OpenCL command submission right away. (With the new runtime hipSYCL is moving in the direction of lazy submission).